Framework/HistogramData/src/Interpolate.cpp

```diff
@@ -0,0 +1,209 @@
+#include "MantidHistogramData/Interpolate.h"
+#include "MantidHistogramData/Histogram.h"
+
+#include <gsl/gsl_errno.h>
+#include <gsl/gsl_spline.h>
+
+#include <sstream>
+
+using Mantid::HistogramData::Histogram;
+using Mantid::HistogramData::HistogramX;
+using Mantid::HistogramData::HistogramY;
+
+namespace {
+
+constexpr const char *LINEAR_NAME = "Linear";
+constexpr const char *CSPLINE_NAME = "CSpline";
+
+/**
+ * Compute the number of pre-calculated points given the ysize and step size
+ * @param ysize The total number of points in the data set
+ * @param stepSize The step size between each calculated point
+ * @return The number of calculated nodes
+ */
+inline size_t numberCalculated(const size_t ysize, const size_t stepSize) {
+  // First and last points are always assumed to be calculated
+  return 1 + ((ysize - 1) / stepSize);
+}
+
+/**
+ * Perform common sanity checks for interpolations
+ * @param input See interpolateLinear
+ * @param stepSize See interpolateLinear
+ * @param minCalculated The minimum number of calculated values required
+ * by the routine
+ */
+void sanityCheck(const Histogram &input, const size_t stepSize,
+                 const size_t minCalculated, const char *method) {
+  if (input.yMode() == Histogram::YMode::Uninitialized) {
+    throw std::runtime_error(
+        "interpolate - YMode must be defined for input histogram.");
+  }
+  const auto ysize = input.y().size();
+  if (stepSize >= ysize) {
+    throw std::runtime_error("interpolate - Step size must be smaller "
+                             "than the number of points");
+  }
+  // First and last points are always assumed to be calculated
+  const size_t ncalc = numberCalculated(ysize, stepSize);
+  if (ncalc < minCalculated) {
+    std::ostringstream os;
+    os << "interpolate - " << method << " requires " << minCalculated
+       << " points but only " << ncalc << " were found.";
+    throw std::runtime_error(os.str());
+  }
+}
+
+/**
+ * Perform linear interpolation assuming the input data set to be a Histogram
+ * with XMode=Points. It is assumed all sanity cheks have been performed
+ * by the interpolateLinear entry point.
+ * @param input See interpolateLinear
+ * @param stepSize See interpolateLinear
+ * @param ynew A reference to the output Y values
+ * @return See interpolateLinear
+ */
+void interpolateYLinearInplace(const Histogram &input, const size_t stepSize,
+                               HistogramY &ynew) {
+  auto xold = input.points();
+  auto &yold = input.y();
```

Comment:
`input` itself is const so `auto &yold` will remain `const` but `xold` should be `const`